File: app/src/main/java/com/mtime/player/MTimeDataProvider.java
package com.mtime.player;

import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;

import com.kk.taurus.playerbase.entity.DataSource;
import com.kk.taurus.playerbase.event.BundlePool;
import com.kk.taurus.playerbase.event.EventKey;
import com.kk.taurus.playerbase.provider.BaseDataProvider;
import com.kotlin.android.app.data.entity.video.VideoPlayList;
import com.kotlin.android.app.data.entity.video.VideoPlayUrl;
import com.kotlin.android.audio.floatview.component.aduiofloat.constant.AudioConstantKt;
import com.mtime.base.network.NetworkException;
import com.mtime.base.network.NetworkManager;
import com.mtime.player.bean.SchemeType;
import com.mtime.bussiness.video.api.MovieApi;
import com.mtime.bussiness.video.bean.VideoInfoApiRequestBean;
import com.mtime.bussiness.video.bean.VideoInfoBean;
import com.mtime.player.bean.MTimeVideoData;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by mtime on 2017/10/19.
 */

public class MTimeDataProvider extends BaseDataProvider {

    private final String TAG = "MTimeDataProvider";

    private final String API_TAG_GET_PLAY_URL = "player_get_play_url";
    private final String API_TAG_POST_PLAY_VIDEO_INFO = "player_get_play_video_info";
    private final MovieApi mMovieApi;
    private boolean cancelFlag;

    public MTimeDataProvider() {
        mMovieApi = new MovieApi();
    }

    @Override
    public void handleSourceData(DataSource data) {
        if (data == null)
            return;
        cancelFlag = false;
        AudioConstantKt.sendCloseEvent();
        if (!TextUtils.isEmpty(data.getData())) {
            Bundle obtain = BundlePool.obtain();
            obtain.putSerializable(EventKey.SERIALIZABLE_DATA, data);
            onProviderMediaDataSuccess(obtain);
            return;
        }
        if (data instanceof MTimeVideoData) {
            handleGetVideoInfo((MTimeVideoData) data);
        }
    }

    /**
     * 获取视频基础信息
     *
     * @param data
     */
    private void handleGetVideoInfo(MTimeVideoData data) {
        //callback start
        onProviderDataStart();
        handleUrlGet(data);
//        VideoInfoApiRequestBean videoInfoApiRequestBean = new VideoInfoApiRequestBean();
//        List<VideoInfoApiRequestBean.VideoInfoApiRequestItem> videoInfoApiRequestItems = new ArrayList<>();
//        VideoInfoApiRequestBean.VideoInfoApiRequestItem requestItem = new VideoInfoApiRequestBean.VideoInfoApiRequestItem();
//        requestItem.setVideoId(Long.parseLong(data.getVideoId()));
//        requestItem.setSource(data.getSource());
//        videoInfoApiRequestItems.add(requestItem);
//        videoInfoApiRequestBean.setScheme(SchemeType.HTTP.getType());
//        videoInfoApiRequestBean.setVideoInfoApiRequest(videoInfoApiRequestItems);
//        mMovieApi.cancelRequest(API_TAG_POST_PLAY_VIDEO_INFO);
//        mMovieApi.getVideoInfo(API_TAG_POST_PLAY_VIDEO_INFO, videoInfoApiRequestBean, new NetworkManager.NetworkListener<VideoInfoBean>() {
//            @Override
//            public void onSuccess(VideoInfoBean result, String showMsg) {
//                if (result != null) {
//                    Bundle bundle = new Bundle();
//                    bundle.putSerializable(EventKey.SERIALIZABLE_DATA, result);
//                    onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_VIDEO_INFO_READY, bundle);
//                    if (result.isPlayable(data.getVideoId())) {
//                        handleUrlGet(data);
//                    } else {
//                        onProviderMediaDataError(null);
//                    }
//                }
//            }
//
//            @Override
//            public void onFailure(NetworkException<VideoInfoBean> exception, String showMsg) {
//                Log.d(TAG, "showMsg = " + showMsg);
//                onProviderMediaDataError(null);
//            }
//        });
    }

    /**
     * 获取视频各清晰度的播放链接
     *
     * @param data
     */
    private void handleUrlGet(final MTimeVideoData data) {
        if (cancelFlag)
            return;
        mMovieApi.cancelRequest(API_TAG_GET_PLAY_URL);
        mMovieApi.getPlayUrl(API_TAG_GET_PLAY_URL, data.getVideoId(), data.getSource(), SchemeType.HTTP.getType(),
                new NetworkManager.NetworkListener<VideoPlayList>() {
                    @Override
                    public void onSuccess(VideoPlayList result, String showMsg) {
                        if (result != null && !cancelFlag) {
                            List<VideoPlayUrl> urlItems = result.getPlayUrlList();
                            if (urlItems != null && urlItems.size() > 0) {
                                handleGetUrlResponseData(data, urlItems);
                                //send get video info event
                                Bundle bundle = new Bundle();
                                bundle.putSerializable(DataInter.Key.KEY_PROVIDER_PLAY_URL_INFO, result);
                                onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_VIDEO_RATE_DATA, bundle);
                            }
                        }
                    }

                    @Override
                    public void onFailure(NetworkException<VideoPlayList> exception, String showMsg) {
                        onProviderMediaDataError(null);
                    }
                });
    }

    /**
     * 处理响应的链接数据
     *
     * @param data
     * @param urlItems
     */
    private void handleGetUrlResponseData(MTimeVideoData data, List<VideoPlayUrl> urlItems) {
        int size = urlItems.size();
        //默认使用最高清晰度起播。
        String recordDefinition = PlayerHelper.getHistoryDefinition();
        VideoPlayUrl playItem = PlayerHelper.getHighDefinition(urlItems);
        for (int i = 0; i < size; i++) {
            VideoPlayUrl item = urlItems.get(i);
            //如果本地有清晰度记录，就以记录清晰度起播
            if (String.valueOf(item.getResolutionType()).equalsIgnoreCase(recordDefinition)) {
                playItem = item;
            }
        }

        //处理最终的播放数据
        handleDataSourceReady(data, playItem);
    }

    /**
     * 处理最终播放数据，回调给播放器
     *
     * @param data
     * @param playItem
     */
    private void handleDataSourceReady(MTimeVideoData data,VideoPlayUrl playItem) {
        MTimeVideoData playItemData = new MTimeVideoData(playItem.getUrl());
        playItemData.setVideoId(data.getVideoId());
        playItemData.setSource(data.getSource());
        //set tag for definition
        playItemData.setTag(playItem.getName());
        playItemData.setFileSize(playItem.getFileSize());

        //将源数据的起播时间点设置上
        playItemData.setStartPos(data.getStartPos());
        onDataSourceReady(playItemData);
    }

    protected void onDataSourceReady(DataSource data) {
        Bundle bundle = BundlePool.obtain();
        bundle.putSerializable(EventKey.SERIALIZABLE_DATA, data);
        onProviderMediaDataSuccess(bundle);
    }

    @Override
    public void cancel() {
        cancelFlag = true;
        mMovieApi.cancelAllTags();
    }

    @Override
    public void destroy() {
        cancel();
    }
}

File: player/src/main/java/com/kotlin/android/player/dataprovider/MTimeDataProvider.kt
package com.kotlin.android.player.dataprovider

import android.os.Bundle
import android.text.TextUtils
import com.kk.taurus.playerbase.config.PConst
import com.kk.taurus.playerbase.entity.DataSource
import com.kk.taurus.playerbase.event.BundlePool
import com.kk.taurus.playerbase.event.EventKey
import com.kk.taurus.playerbase.provider.BaseDataProvider
import com.kk.taurus.playerbase.utils.NetworkUtils
import com.kotlin.android.core.CoreApp
import com.kotlin.android.app.data.entity.live.CameraPlayUrl
import com.kotlin.android.app.data.entity.live.DanmuBean
import com.kotlin.android.app.data.entity.live.DirectorUnits
import com.kotlin.android.app.data.entity.video.VideoDetail
import com.kotlin.android.app.data.entity.video.VideoPlayList
import com.kotlin.android.app.data.entity.video.VideoPlayUrl
import com.kotlin.android.audio.floatview.component.aduiofloat.constant.AUDIO_ACTION_CLOSE
import com.kotlin.android.audio.floatview.component.aduiofloat.constant.postEvent
import com.kotlin.android.audio.floatview.component.aduiofloat.constant.sendCloseEvent
import com.kotlin.android.audio.floatview.component.aduiofloat.event.EventAudioPlayer
import com.kotlin.android.ktx.ext.log.e
import com.kotlin.android.player.DataInter
import com.kotlin.android.player.PlayerConfig
import com.kotlin.android.player.PlayerHelper
import com.kotlin.android.player.bean.LiveVideoData
import com.kotlin.android.player.bean.MTimeVideoData
import com.kotlin.android.player.splayer.PreviewVideoPlayer

/**
 * create by lushan on 2020/9/3
 * description:视频播放数据provider,提供视频播放地址、清晰度和视频文件大小
 */
class MTimeDataProvider(var listener: ((MTimeVideoData) -> Unit)? = null) : BaseDataProvider() {
    companion object {
        const val KEY_IS_LIVE = "is_live"
    }

    private var mDataSource = MTimeVideoData("", 0L, 0L, 0L)
    override fun handleSourceData(sourceData: DataSource?) {
        sourceData ?: return

        sendCloseEvent()//关闭音频
        "handleSourceData==".e()
        if (!TextUtils.isEmpty(sourceData.data)) {
            val obtain = BundlePool.obtain()
            obtain.putSerializable(EventKey.SERIALIZABLE_DATA, sourceData)
            onProviderMediaDataSuccess(obtain)
            return
        }
        if (sourceData is MTimeVideoData) {
            "handleSourceData==加载视频播放地址".e()
            mDataSource = sourceData
            listener?.invoke(sourceData)
        }
    }

    /**
     * 当前播放视频的视频id
     */
    fun getCurrentVideoId(): Long {
        return mDataSource.videoId
    }

    /**
     * 更新视频id和视频类型
     */
    fun updateSourceData(videoId: Long, sourceType: Long) {
        mDataSource.videoId = videoId
        mDataSource.source = sourceType
    }

    /**
     * 设置视频播放地址
     */
    fun setVideoPlayUrlList(bean: VideoPlayList,notIgnoreMobileNet:Boolean = true) {
        val list = bean.playUrlList ?: arrayListOf()
//        "$list".e()
        if (list.isNotEmpty()) {
            //默认使用最高清晰度起播。
            val recordDefinition: String = PlayerHelper.getHistoryDefinition().orEmpty()
            var playItem: VideoPlayUrl? = PlayerHelper.getHighDefinition(list)
            list.forEach {
                //如果本地有清晰度记录，就以记录清晰度起播
                if (it.resolutionType.toString().equals(recordDefinition, ignoreCase = true)) {
                    playItem = it
                }
            }

            mDataSource.isLive = false

            //处理最终的播放数据
            handleDataSourceReady(mDataSource, playItem)
            val bundle = Bundle()
            bundle.putSerializable(DataInter.Key.KEY_PROVIDER_PLAY_URL_INFO, bean)
            onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_VIDEO_RATE_DATA, bundle)
            if (notIgnoreMobileNet){
                stopWithMobileNet()
            }else{
                PlayerConfig.ignoreMobile = true
            }
        }
    }

    private fun stopWithMobileNet() {
        if (NetworkUtils.getNetworkState(CoreApp.instance.applicationContext) > PConst.NETWORK_STATE_WIFI && !PlayerConfig.ignoreMobile) {
            PreviewVideoPlayer.get()?.requestStop()
            onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_VIDEO_INFO_READY, null)

        }
    }


    /**
     * 处理直播视频播放地址
     */
    fun setLiveVideoPlayUrlList(bean: CameraPlayUrl?) {
        bean ?: return
        if (bean.hd == null && bean.ori == null && bean.sd == null) {
            return
        }
        val videoPlayerList: ArrayList<VideoPlayUrl> = arrayListOf()
        val recordDefinition: String = PlayerHelper.getHistoryDefinition().orEmpty()
        var playItem: VideoPlayUrl? = null
        var isLive = true
        bean.sd?.apply {
            videoPlayerList.add(
                VideoPlayUrl(
                    url = this.rtmpUrl.orEmpty(),
                    name = "标清",
                    resolutionType = 1L,
                    isLive = isLive
                )
            )
        }
        bean.hd?.apply {
            playItem = VideoPlayUrl(
                url = this.rtmpUrl.orEmpty(),
                name = "高清",
                resolutionType = 2L,
                isLive = isLive
            )
            playItem?.apply {
                videoPlayerList.add(this)
            }
        }
        bean.ori?.apply {
            videoPlayerList.add(
                VideoPlayUrl(
                    url = this.rtmpUrl.orEmpty(),
                    name = "原画",
                    resolutionType = 3L,
                    isLive = isLive
                )
            )
        }
        videoPlayerList.forEach {
            //如果本地有清晰度记录，就以记录清晰度起播
            if (it.resolutionType.toString().equals(recordDefinition, ignoreCase = true)) {
                playItem = it
            }
        }
        mDataSource.isLive = isLive
        //处理最终的播放数据
        handleDataSourceReady(mDataSource, playItem)
        val bundle = Bundle()
        bundle.putSerializable(
            DataInter.Key.KEY_PROVIDER_PLAY_URL_INFO,
            VideoPlayList(videoPlayerList)
        )
        onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_VIDEO_RATE_DATA, bundle)
        refreshCameraStandList()
//        LogUtils.e("机位地址获取","setLiveVideoPlayUrlList机位播放地址获取成功判断网络1")
        stopWithMobileNet()
    }


    /**
     * 视频播放地址加载失败
     */
    fun setVideoPlayUrlError() {
        onProviderMediaDataError(Bundle())
        onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_VIDEO_NO_RATE_DATA, Bundle())
    }

    private fun handleDataSourceReady(dataSource: MTimeVideoData, playItem: VideoPlayUrl?) {
//        "handleDataSourceReady加载后播放视频:${playItem?.url}".e()
        val mTimeVideoData = MTimeVideoData(
            playItem?.url.orEmpty(), dataSource.videoId, dataSource.source, playItem?.fileSize
                ?: 0L
        )
        mTimeVideoData.isLive = dataSource.isLive
        mTimeVideoData.tag = playItem?.name
        mTimeVideoData.startPos = dataSource.startPos
        onDataSourceReady(mTimeVideoData)
    }

    fun refreshVideoDetail(detail: VideoDetail?) {
        detail?.apply {
            val bundle = Bundle()
            bundle.putSerializable(EventKey.SERIALIZABLE_DATA, this)
            onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_VIDEO_INFO_READY, bundle)
        }
    }

    /**
     * 刷新直播视频观看人数、直播标识
     */
    fun refreshLiveVideoDetail(detail: LiveVideoData?) {
        detail?.apply {
            val bundle = Bundle()
            bundle.putSerializable(EventKey.SERIALIZABLE_DATA, this)
            onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_VIDEO_INFO_READY, bundle)
        }
    }

    /**
     * 刷新机位列表
     */
    fun refreshCameraStandList() {
        //        显示直播机位列表
        onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_LIVE_CAMERA_LIST, null)
    }

    /**
     * 根据横竖屏刷新机位
     */
    fun refreshCameraStandUIByOrientation(isPortrait: Boolean) {
        Bundle().apply {
            putBoolean(DataInter.Key.KEY_LIVE_CAMERA_LIST_ORIENTATION, isPortrait)
        }.also {
            onProviderExtraDataSuccess(
                DataInter.ProviderEvent.EVENT_LIVE_CAMERA_LIST_ORIENTATION,
                it
            )
        }
    }

    private fun onDataSourceReady(data: MTimeVideoData) {
        val bundle = BundlePool.obtain()
        bundle.putSerializable(EventKey.SERIALIZABLE_DATA, data)
        onProviderMediaDataSuccess(bundle)
    }

    /**
     * 更新直播发送消息
     */
    fun updateChatContent(content: String?) {
        Bundle().apply {
            putString(DataInter.Key.KEY_LIVE_SEND_CHAT_CONTENT, content.orEmpty())
        }.also {
            onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_LIVE_SEND_CHAT_CONTENT, it)
        }
    }

    /**
     * 更新弹幕集合
     */
    fun updateDanmuList() {
        onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_LIVE_DANMU, null)
    }

    /**
     * 更新单个弹幕
     */
    fun updateDanmu(bean: DanmuBean) {
        Bundle().apply {
            putSerializable(DataInter.Key.KEY_LIVE_DANMU_CONTENT, bean)
        }.also {
            onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_LIVE_DANMU, it)
        }
    }

    /**
     * 更新导播台
     */
    fun updateDirectorUnits(bean:DirectorUnits){
        Bundle().apply {
            putSerializable(DataInter.Key.KEY_LIVE_DIRECTOR_UNITS, bean)
        }.also {
            onProviderExtraDataSuccess(DataInter.ProviderEvent.EVENT_LIVE_DIRECTOR_UNITS, it)
        }
    }

    override fun cancel() {}

    override fun destroy() {}
}